File: k9mail/src/main/java/com/fsck/k9/autocrypt/AutocryptHeader.java
package com.fsck.k9.autocrypt;


import java.util.Arrays;
import java.util.Map;

import android.support.annotation.NonNull;

import okio.ByteString;


class AutocryptHeader {
    static final String AUTOCRYPT_HEADER = "Autocrypt";

    static final String AUTOCRYPT_PARAM_ADDR = "addr";
    static final String AUTOCRYPT_PARAM_KEY_DATA = "keydata";

    static final String AUTOCRYPT_PARAM_TYPE = "type";
    static final String AUTOCRYPT_TYPE_1 = "1";

    static final String AUTOCRYPT_PARAM_PREFER_ENCRYPT = "prefer-encrypt";
    static final String AUTOCRYPT_PREFER_ENCRYPT_MUTUAL = "mutual";

    private static final int HEADER_LINE_LENGTH = 76;


    @NonNull
    final byte[] keyData;
    @NonNull
    final String addr;
    @NonNull
    final Map<String,String> parameters;
    final boolean isPreferEncryptMutual;

    AutocryptHeader(@NonNull Map<String, String> parameters, @NonNull String addr,
            @NonNull byte[] keyData, boolean isPreferEncryptMutual) {
        this.parameters = parameters;
        this.addr = addr;
        this.keyData = keyData;
        this.isPreferEncryptMutual = isPreferEncryptMutual;
    }

    String toRawHeaderString() {
        // TODO we don't properly fold lines here. if we want to support parameters, we need to do that somehow
        if (!parameters.isEmpty()) {
            throw new UnsupportedOperationException("arbitrary parameters not supported");
        }

        StringBuilder builder = new StringBuilder();
        builder.append(AutocryptHeader.AUTOCRYPT_HEADER).append(": ");
        builder.append(AutocryptHeader.AUTOCRYPT_PARAM_ADDR).append('=').append(addr).append("; ");
        if (isPreferEncryptMutual) {
            builder.append(AutocryptHeader.AUTOCRYPT_PARAM_PREFER_ENCRYPT)
                    .append('=').append(AutocryptHeader.AUTOCRYPT_PREFER_ENCRYPT_MUTUAL).append("; ");
        }
        builder.append(AutocryptHeader.AUTOCRYPT_PARAM_KEY_DATA).append("=");

        appendBase64KeyData(builder);

        return builder.toString();
    }

    private void appendBase64KeyData(StringBuilder builder) {
        String base64KeyData = ByteString.of(keyData).base64();

        int base64Length = base64KeyData.length();
        int lineLengthBeforeKeyData = builder.length();
        int dataLengthInFirstLine = HEADER_LINE_LENGTH -lineLengthBeforeKeyData;

        boolean keyDataFitsInFirstLine = dataLengthInFirstLine > 0 && base64Length < dataLengthInFirstLine;
        if (keyDataFitsInFirstLine) {
            builder.append(base64KeyData, 0, base64Length);
            return;
        }

        if (dataLengthInFirstLine > 0) {
            builder.append(base64KeyData, 0, dataLengthInFirstLine).append("\r\n ");
        } else {
            builder.append("\r\n ");
            dataLengthInFirstLine = 0;
        }

        for (int i = dataLengthInFirstLine; i < base64Length; i += HEADER_LINE_LENGTH) {
            if (i + HEADER_LINE_LENGTH <= base64Length) {
                builder.append(base64KeyData, i, i + HEADER_LINE_LENGTH).append("\r\n ");
            } else {
                builder.append(base64KeyData, i, base64Length);
            }
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        AutocryptHeader that = (AutocryptHeader) o;

        if (isPreferEncryptMutual != that.isPreferEncryptMutual) {
            return false;
        }
        if (!Arrays.equals(keyData, that.keyData)) {
            return false;
        }
        if (!addr.equals(that.addr)) {
            return false;
        }
        if (!parameters.equals(that.parameters)) {
            return false;
        }

        return true;
    }

    @Override
    public int hashCode() {
        int result = Arrays.hashCode(keyData);
        result = 31 * result + addr.hashCode();
        result = 31 * result + parameters.hashCode();
        result = 31 * result + (isPreferEncryptMutual ? 1 : 0);
        return result;
    }
}

File: legacy/core/src/main/java/com/fsck/k9/autocrypt/AutocryptHeader.kt
package com.fsck.k9.autocrypt

import okio.ByteString

data class AutocryptHeader(
    val parameters: Map<String, String>,
    val addr: String,
    val keyData: ByteArray,
    val isPreferEncryptMutual: Boolean,
) {

    fun toRawHeaderString(): String {
        // TODO we don't properly fold lines here. if we want to support parameters, we need to do that somehow
        if (parameters.isNotEmpty()) {
            error("arbitrary parameters not supported")
        }

        return buildString {
            append(AUTOCRYPT_HEADER).append(": ")
            append(AUTOCRYPT_PARAM_ADDR).append('=').append(addr).append("; ")
            if (isPreferEncryptMutual) {
                append(AUTOCRYPT_PARAM_PREFER_ENCRYPT)
                    .append('=').append(AUTOCRYPT_PREFER_ENCRYPT_MUTUAL).append("; ")
            }
            append(AUTOCRYPT_PARAM_KEY_DATA).append("=")
            append(createFoldedBase64KeyData(keyData))
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is AutocryptHeader) return false

        return isPreferEncryptMutual == other.isPreferEncryptMutual &&
            keyData.contentEquals(other.keyData) &&
            addr == other.addr &&
            parameters == other.parameters
    }

    override fun hashCode(): Int {
        var result = keyData.contentHashCode()
        result = 31 * result + addr.hashCode()
        result = 31 * result + parameters.hashCode()
        result = 31 * result + if (isPreferEncryptMutual) 1 else 0
        return result
    }

    companion object {
        const val AUTOCRYPT_HEADER = "Autocrypt"

        const val AUTOCRYPT_PARAM_ADDR = "addr"
        const val AUTOCRYPT_PARAM_KEY_DATA = "keydata"

        const val AUTOCRYPT_PARAM_TYPE = "type"
        const val AUTOCRYPT_TYPE_1 = "1"

        const val AUTOCRYPT_PARAM_PREFER_ENCRYPT = "prefer-encrypt"
        const val AUTOCRYPT_PREFER_ENCRYPT_MUTUAL = "mutual"

        private const val HEADER_LINE_LENGTH = 76

        fun createFoldedBase64KeyData(keyData: ByteArray): String {
            val base64KeyData = ByteString.of(data = keyData).base64()
            val result = StringBuilder()

            for (i in base64KeyData.indices step HEADER_LINE_LENGTH) {
                result.append("\r\n ")
                if (i + HEADER_LINE_LENGTH <= base64KeyData.length) {
                    result.append(base64KeyData, i, i + HEADER_LINE_LENGTH)
                } else {
                    result.append(base64KeyData, i, base64KeyData.length)
                }
            }

            return result.toString()
        }
    }
}
